source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,330 @@
+ *
+ */
+
+// Template for integer types.
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < sizeof(T)) {
+      return false;
+    }
+    value_ = data.peekBEInt<T>(pos);
+    pos += sizeof(T);
+    left -= sizeof(T);
+    return true;
+  }
+
+  std::string toString() const {
+    char buf[32];
+    sprintf(buf, getFormat(), value_);
+    return std::string(buf);
+  }
+
+  constexpr const char* getFormat() const { return "[%02d]"; }
+  T get() const { return value_; }
+
+private:
+  T value_{};
+};
+
+using Int32 = Int<uint32_t>;
+using Int16 = Int<uint16_t>;
+using Int8 = Int<uint8_t>;
+
+// 8-bits character value
+using Byte1 = Int<char>;
+
+// getFormat method specialization for Byte1.
+template <> constexpr const char* Int<char>::getFormat() const { return "[%c]"; }
+
+// String type requires byte with zero value to indicate end of string.
+class String {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
+    auto index = data.search(&zero, 1, pos);
+    if (index == -1) {
+      return false;
+    }
+
+    // Reserve that many bytes in the string.
+    auto size = index - pos;
+    value_.resize(size);
+    // Now copy from buffer to string.
+    data.copyOut(pos, index - pos, value_.data());
+    pos += (size + 1);
+    left -= (size + 1);
+
+    return true;
+  }
+
+  std::string toString() const { return absl::StrCat("[", value_, "]"); }
+
+private:
+  std::string value_;
+};
+
+// ByteN type is used as the last type in the Postgres message and contains
+// sequence of bytes. The length must be deduced from message length.
+class ByteN {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if (left > (data.length() - pos)) {
+      return false;
+    }
+    value_.resize(left);
+    data.copyOut(pos, left, value_.data());
+    pos += left;
+    left = 0;
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out = "[";
+    out.reserve(value_.size() * 3);
+    bool first = true;
+    for (const auto& i : value_) {
+      char buf[16];
+      if (first) {
+        sprintf(buf, "%02d", i);
+        first = false;
+      } else {
+        sprintf(buf, " %02d", i);
+      }
+      absl::StrAppend(&out, buf);
+    }
+    absl::StrAppend(&out, "]");
+    return out;
+  }
+
+private:
+  std::vector<uint8_t> value_;
+};
+
+// Class represents the structure consisting of 4 bytes of length
+// indicating how many bytes follow.
+// In Postgres documentation it is described as:
+// Int32
+// The length of the function result value, in bytes (this count does not include itself). Can be
+// zero. As a special case, -1 indicates a NULL function result. No value bytes follow in the NULL
+// case.
+//
+// ByteN
```

Comment:
These should be bulleted points?

Reply:
Changed entire comment to be more descriptive.